File: 32-longest-valid-parentheses/32-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        # check base case
        if not s:
            return 0
        stack = [-1]
        max_len = 0 
        
        for i,ch in enumerate(s):
            if ch == '(':
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    curr_len = i - stack[len(stack) - 1]
                    max_len  = max(curr_len, max_len)
        return max_len
        
        
        
        

File: 32-longest-valid-parentheses/32-longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int MAX = 0; // variable for length of max valid parentheses
        stack<int> stk;
        stk.push(-1); 
        for(int i = 0; i < size(s); i++)
            if(s[i] == '(') stk.push(i);
            else{
                stk.pop();
                if(stk.empty()) stk.push(i);
                else MAX = max(MAX, i - stk.top());
            }
        return MAX;
        
    }
};